src/chain/chain-training.cc

```diff
@@ -103,13 +104,57 @@ void ComputeChainObjfAndDeriv(const ChainTrainingOptions &opts,
     *l2_term = 0.0;
   } else {
     // compute the l2 penalty term and its derivative
-    BaseFloat scale = supervision.weight * opts.l2_regularize;
-    *l2_term = -0.5 * scale * TraceMatMat(nnet_output, nnet_output, kTrans);
-    if (nnet_output_deriv)
-      nnet_output_deriv->AddMat(-1.0 * scale, nnet_output);
+    BaseFloat scale_coeff = supervision.weight * opts.l2_regularize;
+    // If xent_output provided, l2 penalty is trying to regress the chain output
+    // to be a linear function of cross-entropy output.
+    // It minimizes -0.5 * l2_regularize * l2_norm(diag(scale) * x + offset - y)^2, 
+    // where x is cross-entropy output and y is chain output.
+    if (xent_output) {
+      //compute offset and scale
+      // The objecitve is to minimize L w.r.t scale_i, offset_i, 
+      // L = -0.5 * l2_regularize * 
+      //    \sum_{j=1}^{minibatch_size}(\sum_i (nnet_output_ji - target_ji)^2),
+      // where the target_ji = scale_i * xent_output_ji + offset_i. 
+      // scale_i = \sum_j (nnet_output_ji * xent_output_ji) / \sum_j(xent_output_ji^2)
```

Comment:
btw, this is linear regression in one dimension.  you could just look up the equations if you want.